bench: add separate-process tenant benchmarks

yuzefovich opened the pull request:
Some time ago we merged a change to run all benchmarks in `pkg/bench` against an in-memory tenant. Recently we introduced a local fastpath for in-process in-memory tenants which will resemble how things will look once we get to UA for single-tenant clusters. However, it is also interesting to measure how we perform with separate-process tenants (which is how we run serverless), so this commit introduces that additional config.

Epic: CRDB-14837

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/95715)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95715)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[`pkg/bench/foreachdb.go` line 101 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95715#-NMX17nl1YCRyyHvhiXB:-NMX17nl1YCRyyHvhiXC:b-i7zd2o) ([raw file](https://github.com/cockroachdb/cockroach/blob/badfb0f7fff854e189fd7e2b5ccc4690f92a37c4/pkg/bench/foreachdb.go#L101)):*
> ```Go
> 	_, tenantDB := serverutils.StartTenant(b, s, base.TestTenantArgs{
> 		TenantName: "benchtenant",
> 		TenantID:   roachpb.MustMakeTenantID(10),
> ```

I initially used `UseDatabase` option with `"cluster:"+tenantName+"/bench"` value, but this made `BenchmarkPgbench` fail. Removing `UseDatabase` seems to work for all benchmarks.
___
*[`pkg/bench/foreachdb.go` line 105 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95715#-NMX1INJ1838PvDDjugx:-NMX1INJ1838PvDDjugy:b-m1q9e3) ([raw file](https://github.com/cockroachdb/cockroach/blob/badfb0f7fff854e189fd7e2b5ccc4690f92a37c4/pkg/bench/foreachdb.go#L105)):*
> ```Go
> 
> 	// The benchmarks sometime hit the default span limit, so we increase it.
> 	// NOTE(andrei): Benchmarks drop the tables they're creating, so I'm not sure
> ```

I haven't tried running the benchmarks without this, so I dunno whether it is still relevant.
___
*[`pkg/bench/bench_test.go` line 1195 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95715#-NMX1TFn90tbsYBRPba-:-NMX1TFn90tbsYBRPba0:b-m0zw10) ([raw file](https://github.com/cockroachdb/cockroach/blob/badfb0f7fff854e189fd7e2b5ccc4690f92a37c4/pkg/bench/bench_test.go#L1195)):*
> ```Go
> 	defer log.Scope(b).Close(b)
> 	ForEachDB(b, func(b *testing.B, db *sqlutils.SQLRunner) {
> 		db.Exec(b, `CREATE TABLE bench.abc (a INT PRIMARY KEY, b INT, c INT, INDEX(b), UNIQUE INDEX(c))`)
> ```

I don't understand why the in-memory tenant cockroach benchmarks didn't fail for the cases where `bench.` prefix was missing - perhaps it's due to explicitly specifying the DB name for `OpenDBConn`? If I did use similar value when starting a separate-process tenant via `UseDatabase` parameter, the benchmarks missing `bench.` prefix still failed :man_shrugging: 


<!-- Sent from Reviewable.io -->

yuzefovich commented:
@knz could you please take a look at this?

knz reviewed (commented):
> Reviewed 2 of 3 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95715)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`pkg/bench/bench_test.go` line 1195 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95715#-NMX1TFn90tbsYBRPba-:-NMkRxMZDUEDHRrwRrq0:bhiykw0) ([raw file](https://github.com/cockroachdb/cockroach/blob/badfb0f7fff854e189fd7e2b5ccc4690f92a37c4/pkg/bench/bench_test.go#L1195)):*
>  perhaps it's due to explicitly specifying the DB name for OpenDBConn? 

yes that would be my understanding as well.

> If I did use similar value when starting a separate-process tenant via UseDatabase parameter, the benchmarks missing bench. prefix still failed 🤷‍♂️

I'm not sure I understand the operation mode. How do you "start a separate-process tenant with a UseDatabase parameter"?
___
*[`pkg/bench/bench_test.go` line 1198 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95715#-NMkS9f5AAFw_luk64H6:-NMkS9f5AAFw_luk64H7:b6os1o) ([raw file](https://github.com/cockroachdb/cockroach/blob/00a69e5578a9d92dd4d9e838f9543f4d09852f6d/pkg/bench/bench_test.go#L1198)):*
> ```Go
> 	defer log.Scope(b).Close(b)
> 	ForEachDB(b, func(b *testing.B, db *sqlutils.SQLRunner) {
> 		db.Exec(b, `CREATE TABLE bench.abc (a INT PRIMARY KEY, b INT, c INT, INDEX(b), UNIQUE INDEX(c))`)
> ```

I'm not sure introducing this `bench.` prefix everywhere is the right way to go. It makes the SQL longer and (slightly) harder to read and update.

 If you can't insert the db name in the connection string, you can execute a `USE` or `SET database` statement at the beginning of the benchmark.


<!-- Sent from Reviewable.io -->

yuzefovich committed:
bench: add separate-process tenant benchmarks

Some time ago we merged a change to run all benchmarks in `pkg/bench`
against an in-memory tenant. Recently we introduced a local fastpath for
in-process in-memory tenants which will resemble how things will look
once we get to UA for single-tenant clusters. However, it is also
interesting to measure how we perform with separate-process tenants
(which is how we run serverless), so this commit introduces that
additional config.

Release note: None

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95715)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[`pkg/bench/bench_test.go` line 1195 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95715#-NMX1TFn90tbsYBRPba-:-NMkW4Ij-Ww56sCqn64Z:b-l9hov7) ([raw file](https://github.com/cockroachdb/cockroach/blob/badfb0f7fff854e189fd7e2b5ccc4690f92a37c4/pkg/bench/bench_test.go#L1195)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

>  perhaps it's due to explicitly specifying the DB name for OpenDBConn? 

yes that would be my understanding as well.

> If I did use similar value when starting a separate-process tenant via UseDatabase parameter, the benchmarks missing bench. prefix still failed 🤷‍♂️

I'm not sure I understand the operation mode. How do you "start a separate-process tenant with a UseDatabase parameter"?
</blockquote></details>

I specified `base.TestTenantArgs.UseDatabase` argument, but only I included `"cluster:"+tenantName+"/bench"` in there. The correct way is to just use `bench` there which is what this commit does. This is probably the way to go. (But it shows that I'm a complete newbie with all of this :) )
___
*[`pkg/bench/bench_test.go` line 1198 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95715#-NMkS9f5AAFw_luk64H6:-NMkW1e84xgS3EmL7Z2n:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/00a69e5578a9d92dd4d9e838f9543f4d09852f6d/pkg/bench/bench_test.go#L1198)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

I'm not sure introducing this `bench.` prefix everywhere is the right way to go. It makes the SQL longer and (slightly) harder to read and update.

 If you can't insert the db name in the connection string, you can execute a `USE` or `SET database` statement at the beginning of the benchmark.
</blockquote></details>

Fixed.


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95715)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`pkg/bench/bench_test.go` line 1195 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95715#-NMX1TFn90tbsYBRPba-:-NMkZ1SH94KNlFnJ3WVm:b-uj4z11) ([raw file](https://github.com/cockroachdb/cockroach/blob/badfb0f7fff854e189fd7e2b5ccc4690f92a37c4/pkg/bench/bench_test.go#L1195)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I specified `base.TestTenantArgs.UseDatabase` argument, but only I included `"cluster:"+tenantName+"/bench"` in there. The correct way is to just use `bench` there which is what this commit does. This is probably the way to go. (But it shows that I'm a complete newbie with all of this :) )
</blockquote></details>

To be honest the idea that TestTenantArgs has a "UseDatabase" field is utterly confusing to me. The database is a property of the _client connection_, not of the tenant! And the way this particular field is plumbed through obscures that fact.

But yes what you did here is correct.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
I'm plenty confused too since I had the same mental model that the DB should be specified on the connection, and not instantiating args to create a tenant.

TFTR!

bors r+
